File: Capitulo-10/K_Vizinhos_Mais_Proximos/Exercicios/Typescript/simple_classification.js
function euclideanDistance(p1, p2) {
    return Math.sqrt(p1.reduce(function (sum, val, i) { return sum + Math.pow(val - p2[i], 2); }, 0));
}
function knn(data, query, k) {
    var distances = data.map(function (point) { return ({
        distance: euclideanDistance(point.features, query),
        label: point.label,
    }); });
    distances.sort(function (a, b) { return a.distance - b.distance; });
    var kNearest = distances.slice(0, k).map(function (d) { return d.label; });
    var counts = {};
    kNearest.forEach(function (label) {
        counts[label] = (counts[label] || 0) + 1;
    });
    return Object.keys(counts).reduce(function (a, b) { return (counts[a] > counts[b] ? a : b); });
}
var data = [
    { features: [1, 2], label: "A" },
    { features: [2, 3], label: "A" },
    { features: [3, 1], label: "B" },
    { features: [6, 5], label: "B" },
    { features: [7, 7], label: "B" },
];
var query = [2, 2];
var k = 3;
console.log(knn(data, query, k));

File: Capitulo-10/K_Vizinhos_Mais_Proximos/Exercicios/Python/simple_classification.py
import math
from collections import Counter


def euclidian_distance(point1, point2):
    return math.sqrt(sum((x - y) ** 2 for x, y in zip(point1, point2)))


def knn(data, query, k):
    distances = []

    for point in data:
        distance = euclidian_distance(point["features"], query)
        distances.append((distance, point["label"]))

    distances.sort(key=lambda x: x[0])
    k_nearest = distances[:k]

    labels = [label for _, label in k_nearest]
    most_common = Counter(labels).most_common(1)

    return most_common[0][0]


# Dataset
data = [
    {"features": [1, 2], "label": "A"},
    {"features": [2, 3], "label": "A"},
    {"features": [3, 1], "label": "B"},
    {"features": [6, 5], "label": "B"},
    {"features": [7, 7], "label": "B"},
]

query = [2, 2]
k = 3

print(knn(data, query, k))
